gnrc_lorawan: add initial support for GNRC based LoRaWAN stack (v2)

jia200x opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR adds a full LoRaWAN stack to RIOT originally based on the work with @fjmolinas in https://github.com/RIOT-OS/RIOT/pull/6645 . Compared to the Semtech pkg, this offers some advantages:
- Integrated to GNRC (uses netif (ifconfig, etc), netreg, etc)
- Tickless MAC (low power friendly)
- Lightweight (~4-5 kb less ROM, ~1.5 kb less RAM with proper GNRC_PKTBUF_SIZE value)
- Asynchronous send and receive.
- Notifications from the MAC layer (sort of WIP :)

Implemented features:
- EU868 region
- Class A
- Join via OTAA and ABP
- Link Check commands

To be implemented
- Indication to know when the bands are free
- Duty Cycle restrictions
- Compliance Fopts
- Class C
- Other regions
- RTC support for true low power support.


This PR adds an IEEE-like MAC with MCPS (Mac Common Part Sublayer) and MLME (Mac Layer Management Entity) for handling data and all kind of management (set keys, request link check, etc). Compared to other MACs this one is request-confirmation based and there's a bidirectional communication between the MAC and higher layers.

Higher layers can:
- Request to send data (the MAC can either send it right now or wait until it's possible to send) end receive an asynchronous confirmation to that Request (e.g couldn't receive ACK from the network server, etc)
- Request a MAC management operation (set a key, join the network, ask for connectivity status, etc). Same as before, this is confirmed asynchronously

MAC layer can:
- Indicate the upper layer that data was received and pass all metadata (pkt, port, rssi, etc)
- Indicate the upper layer about an interesting event (join was successful, a sub-band is free to be used again, etc)

This IEEE MAC layer is very useful in scenarios like Tisch or connection oriented MAC layers.


<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
Check the README.md file under `tests/gnrc_lorawan`
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Waiting for ~#11143~, ~#11136~, ~#11144~ , ~#11736~ 
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

jia200x commented:
(ignore the commit messages, they will be squashed of course :)

miri64 commented on drivers/include/net/netdev.h:
```diff
@@ -250,6 +250,7 @@ typedef enum {
 struct netdev_radio_rx_info {
     int16_t rssi;       /**< RSSI of a received packet in dBm */
     uint8_t lqi;        /**< LQI of a received packet */
+    int8_t snr;         /**< SNR of a received packet in dBm */
```
There is misplaced struct in [`drivers/sx127x/include/sx127x_netdev.h`](https://github.com/RIOT-OS/RIOT/blob/cebea3a23ed8ea101912e5133f6e73b4634fb46c/drivers/sx127x/include/sx127x_netdev.h) that you could move to a more generic position (e.g. `drivers/include/net/netdev/lora.h`) instead of changing the API here.

miri64 replied:
Also note https://github.com/RIOT-OS/RIOT/pull/6797#discussion_r260303912

jia200x replied:
agreed!

jia200x replied:
solved with #11144

miri64 commented on drivers/sx127x/sx127x.c:
```diff
@@ -307,9 +307,10 @@ static void _on_tx_timeout(void *arg)
+    //netdev_t *dev = (netdev_t *) arg;
 
-    dev->event_callback(dev, NETDEV_EVENT_RX_TIMEOUT);
+    //dev->event_callback(dev, NETDEV_EVENT_RX_TIMEOUT);
```
TODO? Please mark the PR as WIP if so.

jia200x replied:
This is actually a bug in upstream. The callback here is called from an xtimer timeout rather than the DIO1 interrupt. I will open a PR with the proper fix.

miri64 replied:
Then I propose to fix the bug in upstream first instead of adding commented out code (and wrong comment style for that matter ;-))

miri64 commented on sys/include/net/gnrc/lorawan/lorawan.h:
```diff
@@ -0,0 +1,167 @@
+#define LORAWAN_STATE_IDLE (0)
+#define LORAWAN_STATE_RX_1 (1)
+#define LORAWAN_STATE_RX_2 (2)
+#define LORAWAN_STATE_TX (3)
```
Here and above: style

miri64 commented:
(needs rebase btw)

jia200x commented:
Not WIP anymore :)

jia200x commented:
(EUIs are in little endian. I will add it to the README)

jia200x commented on sys/shell/commands/sc_gnrc_netif.c:
```diff
@@ -285,10 +313,26 @@ static void _print_netopt(netopt_t opt)
+            printf("otaa");
+            break;
+
+        case NETOPT_LINK_CHECK:
```
note to self: this clearly has the wrong printf message

jia200x commented on drivers/include/net/netdev.h:
```diff
@@ -239,6 +239,12 @@ typedef enum {
     NETDEV_EVENT_CRC_ERROR,                 /**< wrong CRC */
     NETDEV_EVENT_FHSS_CHANGE_CHANNEL,       /**< channel changed */
     NETDEV_EVENT_CAD_DONE,                  /**< channel activity detection done */
+    NETDEV_EVENT_MLME_CONFIRM,
```
This requires comments

jia200x commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -138,7 +138,7 @@ static int _recv(netdev_t *netdev, void *buf, size_t len, void *info)
             }
 
-            netdev_sx127x_lora_packet_info_t *packet_info = info;
+            struct netdev_radio_rx_info *packet_info = info;
```
radio fixes will be moved to another PR

jia200x commented on sys/include/net/lorawan/hdr.h:
```diff
@@ -0,0 +1,165 @@
+/*
```
if needed, I can move this hdr files out of this PR

miri64 commented:
Please rebase and squash. >200 commits are impossible to keep track of.

jia200x commented:
I will move the radio commits out of this PR, and then rebase

jia200x commented:
rebased, squashed and addressed some changes in master

jia200x commented:
(waiting for #11136)

jia200x commented:
there were several fixups, so I rebased again

jia200x commented:
I will rebase on top of #11144 as well

jia200x commented on tests/gnrc_lorawan/Makefile:
```diff
@@ -0,0 +1,64 @@
+CFLAGS += -DLORAMAC_DEFAULT_RX2_DR=LORAMAC_DR_3
+
+# Set default messages to unconfirmable
+CFLAGS += -DLORAMAC_DEFAULT_TX_MODE=LORAMAC_TX_UNCNF
```
The problem is most network servers (e.g TTN) have a restriction in the number of downlink messages including ACKs. For TTN, the budget is 10 messages per 24 hours

fjmolinas replied:
Got it. My comment goes more to the fact that since it is a test application (not an example application), the test should by default be as complete as possible, therefore test confirmed messages (which have the downlink factor). But this can get overwritten anyway, so I don't mind with the budget limit info taken into account.

jia200x replied:
I switched the test to CNF. I will mark this as resolved.

jia200x commented on sys/shell/commands/sc_gnrc_netif.c:
```diff
@@ -811,6 +867,36 @@ static int _netif_set_flag(kernel_pid_t iface, netopt_t opt,
+            break;
+        default:
+            /* Same rationale here */
+            expected_len = LORAMAC_DEVEUI_LEN;
```
because this default case only gets triggered with DEVEUI and APPEUI. Both are 8U long

fjmolinas replied:
Got it!

jia200x commented on sys/shell/commands/sc_gnrc_netif.c:
```diff
@@ -497,13 +551,15 @@ static void _netif_list(kernel_pid_t iface)
                                    line_thresh);
     line_thresh += _LINE_THRESHOLD + 1; /* enforce linebreak after this option */
-    line_thresh = _netif_list_flag(iface, NETOPT_AUTOCCA, "AUTOCCA",
+    line_thresh = _netif_list_flag(iface, NETOPT_AUTOCCA, "AUTOCCA ",
```
I added this space because otherwise all lines are displayed like:
```
AUTOCCAIQ_INVERTRX_SINGLE
```
instead of
```
AUTOCCA IQ_INVERT RX_SINGLE
```

Maybe I can move this change to its own commit

jia200x commented on sys/include/net/gnrc/lorawan/lorawan.h:
```diff
@@ -0,0 +1,173 @@
+#define GNRC_LORAWAN_REQ_STATUS_DEFERRED (1)    /**< the MLME or MCPS confirm message is asynchronous */
+
+/**
+ * @brief MAC (sub) Layer Management Entity request representation
```
you mean in all mentions of MLME and MCPS?

jia200x replied:
ah, got it!

jia200x commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan.c:
```diff
@@ -0,0 +1,316 @@
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _DRIFT_FACTOR (1000000 * 100 / (100 + CONFIG_GNRC_LORAWAN_TIMER_DRIFT))
```
I will add a comment explaining the rationale

jia200x replied:
done

jia200x commented:
@fjmolinas thanks for the review!

jia200x commented on tests/gnrc_lorawan/Makefile:
```diff
@@ -0,0 +1,64 @@
+CFLAGS += -DLORAMAC_DEV_EUI_DEFAULT=\{0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\}
+
+# Uncomment and replace with proper keys for joining with ABP
+#CFLAGS += -DLORAMAC_DEV_ADDR_DEFAULT=\{0x00\,0x00\,0x00\,0x00\}
```
With this I meant: If someone wants to use ABP, comment OTA lines and uncomment these ones. I will add a better explanation

jia200x commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan.c:
```diff
@@ -0,0 +1,316 @@
+
+static inline void _set_rx2_dr(gnrc_lorawan_t *mac, uint8_t rx2_dr)
+{
+    mac->dl_settings &= 0xF0;
```
indeed. I moved this file to #11143

jia200x replied:
Fixed

jia200x commented:
rebased

jia200x commented:
rebased again. @fjmolinas @smlng @aabadie @miri64 mind giving it a look? :)

jia200x commented:
@miri64 @kaspar030 Murdock keeps failing in the `gnrc_lorawan` test. I can compile it with the Docker image without any issues.

Any idea?

smlng commented on sys/include/net/gnrc/lorawan.h:
```diff
@@ -0,0 +1,251 @@
+ * @brief MAC Information Base descriptor for MLME Request-Confirm
+ */
+typedef struct {
+    mlme_mib_type_t type; /**< MIB attribute identifier */
```
where do the following struct definitions come from? Because they look a bit inconsistent in style, i.e., some use an `enum` for `type` other generic `uint8_t`. Also the order is different, i.e., `param` after `type` or before. If possible, this should be made consistent.

jia200x replied:
I unified them as enums. See last commits

smlng commented on sys/include/net/gnrc/lorawan.h:
```diff
@@ -0,0 +1,251 @@
+typedef struct {
+    mlme_mib_type_t type; /**< MIB attribute identifier */
+    union {
+        int activation;
```
also using a named `union` with only a single attribute seams unnecessary. Why not use anonymous `union` in all structs instead?

jia200x replied:
This list is intended to be expanded. I will add comments regarding that.
I will check how it goes if I make unions out of all structs

smlng replied:
you can stick with the union(s), but I would not name them to avoid long, complex naming when accessing the values, i.e. `foo->param.activation` could be `foo->activation` when using anonymous unions instead.

jia200x replied:
done!

smlng commented on sys/include/net/gnrc/lorawan.h:
```diff
@@ -0,0 +1,251 @@
+ * @param[in] nwkskey buffer to store the NwkSKey. Should be at least 16 bytes long
+ * @param[in] appskey buffer to store the AppsKey. Should be at least 16 bytes long
+ */
+void gnrc_lorawan_init(gnrc_lorawan_t *mac, uint8_t *nwkskey, uint8_t *appskey);
```
can (some) parameters be made `const`? Check for all functions please.

jia200x replied:
yes, most of them. I will check

jia200x replied:
done! Please mark as reviewed if OK

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -0,0 +1,341 @@
+    }
+}
+
+void gnrc_lorawan_mcps_request(gnrc_lorawan_t *mac, mcps_request_t *mcps_request, mcps_confirm_t *mcps_confirm)
```
to me it looks like you can omit the `mcps_confirm_t` param and make this function `int` (instead of `void`) and return the status code instead of using `mcps_confirm->status`. Because the only usage of this function is [here](https://github.com/RIOT-OS/RIOT/pull/11022/files#diff-385d2a03ea8a97a25dbb4c48d127beb1R210) where you pass in an empty `mcps_confirm_t` just to use its status variable for return afterwards.

jia200x replied:
the problem with that is MCPS and MLME confirm can return other attributes (besides status, Link Check stats, MIB params, etc). The current implementation still doesn't add too many MCPS and MLME confirmations, so it's not that visible yet

jia200x replied:
are you OK with this statement?

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,354 @@
+    gnrc_pktsnip_t *pkt = gnrc_pktbuf_add(NULL, NULL, sizeof(lorawan_join_request_t), GNRC_NETTYPE_UNDEF);
+
+    if (!pkt) {
+        goto out;
```
avoid `goto` by rewriting into
```
if (pkt) {
<do stuff>
}
return pkt;
```

or 
```
if (!pkt) {
   return NULL;
}
<do stuff>
return pkt;
```

jia200x replied:
done

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,354 @@
+void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size)
+{
+    if (!fopts || !size) {
+        goto out;
```
use `return` instead, there nothing done at `out`

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,354 @@
+    _buffer_reset(&buf, fopts, size);
+
+    while (buf.index < buf.size) {
+        switch (*(buf.data)) {
```
switch looks like overkill here, `if ... else` looks cleaner (and avoids `goto`):
```
while (buf.index < buf.size) {
    if ((*(buf.data) != GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS) ||
        (_mlme_link_check_ans(mac, &buf) < 0)) {
        break;
    }
}
```

smlng replied:
I believe the code above should do the same as your `switch` + `goto`🤞

jia200x replied:
This list is intended to be expanded, so I will need a switch anyway :(. I removed the goto in the last fixup though

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,354 @@
+{
+    size_t size = 0;
+
+    size += (mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) ?
```
Why not simplify to:

```
return (mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) ?
        _fopts_mlme_link_check_req(mac, buf) : 0;
```

jia200x replied:
this list is intended to be expanded if there are more fopts there. So I wrote it this way to make it easier to expand

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_region.c:
```diff
@@ -0,0 +1,125 @@
+
+uint8_t gnrc_lorawan_rx1_get_dr_offset(uint8_t dr_up, uint8_t dr_offset)
+{
+    return dr_up > dr_offset ? dr_up - dr_offset : 0;
```
please use parentheses for readability 
```
return (dr_up > dr_offset) ? (dr_up - dr_offset) : 0;
```

smlng commented on sys/include/net/gnrc/lorawan/region.h:
```diff
@@ -0,0 +1,65 @@
+#define GNRC_LORAWAN_LC_2 (868300000)   /**< Second default EU868 channel */
+#define GNRC_LORAWAN_LC_3 (868500000)   /**< Third default EU868 channel */
+
+#define GNRC_LORAWAN_DEFAULT_CHANNELS 3 /**< Number of default channels for the current region */
```
this should be named `GNRC_LORAWAN_DEFAULT_CHANNELS_NUMOF` to be consistent with other similar names in RIOT like `I2C_NUMOF` and so on

smlng replied:
and using an array as suggested above this will be:
```
#define GNRC_LORAWAN_DEFAULT_CHANNELS_NUMOF (sizeof(gnrc_lorawan_default_channels)/sizeof(uint32_t))
```

jia200x replied:
done

smlng commented on sys/include/net/gnrc/lorawan/region.h:
```diff
@@ -0,0 +1,65 @@
+extern "C" {
+#endif
+
+#define GNRC_LORAWAN_LC_1 (868100000)   /**< First default EU868 channel */
```
maybe use a array here, which might simplify usage and extension, i.e.:
```
static const uint32_t gnrc_lorawan_default_channels[] = {
    868100000UL,
    868300000UL,
    868500000UL
}
```

jia200x replied:
done

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_region.c:
```diff
@@ -0,0 +1,125 @@
+
+void gnrc_lorawan_channels_init(gnrc_lorawan_t *mac)
+{
+    mac->channel[0] = GNRC_LORAWAN_LC_1;
```
using the array here this can be converted into a loop as well.

smlng replied:
or maybe even put into a single loop like:
```
for (unsigned i = 0; i < GNRC_LORAWAN_MAX_CHANNELS; i++) {
    if (i < GNRC_LORAWAN_DEFAULT_CHANNELS_NUMOF) {
        mac->channel[i] = gnrc_lorawan_default_channels[i];
    }
    else {
        mac->channel[i] = 0;
    }
}
```

which would also avoid an error if `GNRC_LORAWAN_MAX_CHANNELS < GNRC_LORAWAN_DEFAULT_CHANNELS`

jia200x replied:
done

smlng commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,354 @@
+    }
+}
+
+int _fopts_mlme_link_check_req(gnrc_lorawan_t *mac, lorawan_buffer_t *buf)
```
looks like an internal function, so you can just drop the `mac` param here

smlng commented on tests/gnrc_lorawan/Makefile:
```diff
@@ -0,0 +1,71 @@
+USEMODULE += gnrc_neterr
+USEMODULE += gnrc_netreg
+USEMODULE += hashes
+USEMODULE += crypto
```
I guess some of the modules above should be made dependencies of `gnrc_lorawan` and thus are not required here, i.e. `crypto` is not used in the test anywhere

jia200x replied:
fixed

jia200x commented:
@smlng thanks for the review!

jia200x commented:
@smlng done. There are some open comments here and there, but addressed everything I could

jia200x commented:
@smlng It seems anonymous unions/structs are not C99 :(

MrKevinWeiss commented:
release status ping, should I change the milestone?

jia200x commented:
> release status ping, should I change the milestone?

I would say no, there are some chances this get into the release

jia200x commented:
@aabadie thanks for the review!

> This PR should be split is sub PR and reuse existing ones like #9022.

Yes, you are right. I will

> I also noticed that functions regarding lpm storage were not implemented, they should be removed

They are intended to be implemented by the user (they are currently user space callbacks). I guess I could improve the documentation

> Maybe the test application could be directly moved to an example.

I'm fine with it. I will do it.

jia200x commented on sys/include/net/netopt.h:
```diff
@@ -574,6 +574,89 @@ typedef enum {
+    /**
+     * @brief   (uint8_t*) application EUI (8 bytes length)
+     */
+    NETOPT_APPEUI,
```
agreed, I will

jia200x commented on Makefile.dep:
```diff
@@ -154,6 +163,9 @@ ifneq (,$(filter gnrc_netif,$(USEMODULE)))
   ifneq (,$(filter netdev_eth,$(USEMODULE)))
     USEMODULE += gnrc_netif_ethernet
   endif
+  ifneq (,$(filter gnrc_lorawan,$(USEMODULE)))
```
The other snippets don't have a blank line before. Should I add one for all?

aabadie replied:
Ah, no sorry. I missed the sublevel position of this block. You can leave it as it now.

jia200x commented:
I moved the netopts to #11736

jia200x commented on drivers/sx127x/sx127x_netdev.c:
```diff
@@ -351,6 +351,16 @@ static int _get(netdev_t *netdev, netopt_t opt, void *val, size_t max_len)
             *((int16_t*) val) = (int16_t)sx127x_get_tx_power(dev);
             return sizeof(int16_t);
 
+        case NETOPT_SYNCWORD:
```
done in #11736

jia200x commented on sys/Makefile.include:
```diff
@@ -5,6 +5,9 @@ endif
 ifneq (,$(filter gnrc_netif,$(USEMODULE)))
   USEMODULE_INCLUDES += $(RIOTBASE)/sys/net/gnrc/netif/include
 endif
+ifneq (,$(filter gnrc_lorawan,$(USEMODULE)))
```
same as https://github.com/RIOT-OS/RIOT/pull/11022#discussion_r296711845?

aabadie replied:
This one is not nested so no.

jia200x replied:
done!

jia200x commented on sys/auto_init/netif/auto_init_sx127x.c:
```diff
@@ -56,8 +65,13 @@ void auto_init_sx127x(void)
         sx127x_setup(&sx127x_devs[i], &sx127x_params[i]);
+#ifdef MODULE_GNRC_LORAWAN
+        gnrc_netif_lorawan_create(sx127x_stacks[i], SX127X_STACKSIZE, SX127X_PRIO,
+                              "sx127x", (netdev_t *)&sx127x_devs[i]);
```
where? AFAIS it's aligned with the `gnrc_neitf_raw` call

aabadie replied:
It's not aligned with `sx127x_stacks[i]` above.

jia200x replied:
done

jia200x commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan.c:
```diff
@@ -0,0 +1,400 @@
+
+static uint32_t lora_time_on_air(size_t payload_size, uint8_t dr, uint8_t cr)
+{
+    assert(dr <= 6);
```
done. I also added a small comment to the function

jia200x commented on sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_region.c:
```diff
@@ -0,0 +1,125 @@
+#include "net/gnrc/netif.h"
+#include "net/gnrc/lorawan/region.h"
+
+#define GNRC_LORAWAN_NUMOF_DATARATES 6
```
done

jia200x commented:
I moved `tests/gnrc_lorawan` to `examples`. Also, I renamed `link_layer/gnrc_lorawan` folder to `link_layer/lorawan` to be consistent the rest of the GNRC modules

jia200x commented on tests/gnrc_lorawan/Makefile:
```diff
@@ -0,0 +1,61 @@
+
+# GNRC LoRaWAN has been tested in the following boards. Eventually this list will
+# be expanded.
+BOARD_WHITELIST := b-l072z-lrwan1 lobaro-lorabox
```
I will try them. But since I moved the test to the examples folder, I just removed the board whitelist.

jia200x commented on tests/gnrc_lorawan/README.md:
```diff
@@ -0,0 +1,145 @@
+
+## Send data
+
+After join, send data using `send` comand. E.g to send "Hello RIOT!" to port 2:
```
done

aabadie commented on examples/gnrc_lorawan/main.c:
```diff
@@ -0,0 +1,126 @@
+int tx_cmd(int argc, char **argv)
+{
+    gnrc_pktsnip_t *pkt;
+    uint8_t port = LORAWAN_PORT; /* Default: 2 */
```
Why not use a netopt for lorawan tx port ? I think #9022 is adding one for this.

jia200x replied:
it's already there, in this file (line 94)

aabadie commented on sys/auto_init/netif/auto_init_sx127x.c:
```diff
@@ -21,16 +21,25 @@
 #define SX127X_NUMOF        (sizeof(sx127x_params) / sizeof(sx127x_params_t))
+#else
+/* GNRC LoRaWAN only supports one interface so far */
+#define SX127X_NUMOF        (1)
```
I wouldn't not do it this way: just keep the initial define but use an `assert(SX127X_NUMOF == 1)` if `MODULE_GNRC_LORAWAN` is defined, e.g. line 68.

jia200x replied:
you are right. I will change it

jia200x replied:
done!

aabadie commented:
> >    I also noticed that functions regarding lpm storage were not implemented, they should be removed

> They are intended to be implemented by the user (they are currently user space callbacks). I guess I could improve the documentation

I still think this feature should be cut-out from this initial PR and provided in a follow-up with an example implementation (maybe on eeprom).

Also needs rebase

jia200x commented:
> I still think this feature should be cut-out from this initial PR and provided in a follow-up with an example implementation (maybe on eeprom).

Ok, I will remove it from the PR and squash/rebase

jia200x commented:
I will need to squash first to remove that commit

jia200x commented:
> I will need to squash first to remove that commit

@aabadie is it ok?

benpicco commented:
needs a rebase.

tcschmidt commented:
@smlng @miri64 @aabadie - I believe @jia200x addressed the major comments. Are your change requests still up-to-date?

tcschmidt commented:
@MichelRottleuthner what about your field tests - any failures to report back?

MichelRottleuthner commented:
>what about your field tests - any failures to report back?

Until now the stack worked pretty reliable for me. ~3 months mobile outdoor deployment (~30k upstream frames) showed no problems. I'm using different data rates and low power modes together with gnrc_lorawan backup/restore callbacks and everything works as expected.
Only one small thing comes to my mind: It would be better to use a lower datarate as default for the join (or some mechanism to automatically decrease it on failing joins).

miri64 commented:
> The only think I would really like to have the opinion of @miri64 is the fact I added an "ifconfig up" command.
> Is it fine to leave it in this PR or you think it's better to have another one?
> 
> Also, I modeled it using a NETOPT_LINK_CONNECTED, although I'm aware this should be implemented at interface level

I did not look into the code, but given that the PR at the moment has ~4000 lines of additions, I think I prefere to have it in a separate PR. Not only to simplify the review of this PR, but also to provide the integration of this feature for other network interfaces as well, without this PR being required to be merged.

benpicco commented on sys/shell/commands/sc_gnrc_netif.c:
```diff
@@ -144,7 +149,14 @@ static void _set_usage(char *cmd_name)
+         "       * \"deveui\" - sets Device EUI\n"
+         "       * \"dr\" - sets datarate\n"
+         "       * \"rx2_dr\" - sets datarate of RX2 (lorawan)\n"
+         "       * \"nwkskey\" - sets Network Session Key\n"
```
Those and the other additions in this file should be hidden behind `#ifdef MODULE_GNRC_NETIF_CMD_LORA` so they don't fill up the ROM on devices that don't use LoRA.

jia200x replied:
addressed in 0696f18

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,349 @@
+ *
+ * @file
+ * @author  José Ignacio Alamos <jose.alamos@haw-hamburg.de>
+ */
```
doxygen comment can be close here

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,349 @@
+            }
+            if (!gnrc_lorawan_mac_acquire(mac)) {
+                mlme_confirm->status = -EBUSY;
+                return;
```
wouldn't break to the same here?

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,349 @@
+
+            if (mac->mlme.backoff_budget < 0) {
+                mlme_confirm->status = -EDQUOT;
+                break;
```
like this? Or why use return above and break here?

smlng replied:
the function ends right after exiting `switch` anyway

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,349 @@
+    size_t size = 0;
+
+    size += (mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) ?
+            _fopts_mlme_link_check_req(buf) : 0;
```
though I like ternary, I have the feeling that in this case the following is more readable

```
if (mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) {
    size += _fopts_mlme_link_check_req(buf);
}
```

smlng replied:
if you want to optimise than refine whole function to:

```
uint8_t gnrc_lorawan_build_options(gnrc_lorawan_t *mac, lorawan_buffer_t *buf)
{
    return (uint8_t)(mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) ?
           _fopts_mlme_link_check_req(buf) : 0;
}
```

jia200x replied:
I intentionally left this "size" variable there because there will be more fopts. But I will change it to the `if` pattern

jia200x replied:
this an all comments above: addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/a278c06444f9a157fb32daf6740a5837550f35d7

smlng commented on sys/shell/commands/sc_gnrc_netif.c:
```diff
@@ -502,13 +556,15 @@ static void _netif_list(kernel_pid_t iface)
-    line_thresh = _netif_list_flag(iface, NETOPT_CHANNEL_HOP, "CHAN_HOP",
+    line_thresh = _netif_list_flag(iface, NETOPT_CHANNEL_HOP, "CHAN_HOP ",
+                                   line_thresh);
+    line_thresh = _netif_list_flag(iface, NETOPT_OTAA, "OTAA",
```
add space after `OTAA ` to allow for options to follow

jia200x replied:
see 3799b23

jia200x commented:
@MichelRottleuthner I changed the documentation in https://github.com/RIOT-OS/RIOT/pull/11022/commits/041d964ba049dd96039a992754035c75de63c35e for clarifying the DR using during OTAA.
I plan to add a LORAMAC_DEFAULT_JOIN_DR in the future to have more flexibility.

jia200x commented:
I plan to rebase on top of https://github.com/RIOT-OS/RIOT/pull/12603. It should be imminent

miri64 commented:
> I plan to rebase on top of #12603. It should be imminent

IMHO #12603 is ready to merge (except for squashing). You could also wait until it is in master and rebase to master directly.

jia200x commented:
I chose the wrong words: I meant to rebase on top of master after #12603 gets merged :)

miri64 commented:
REady to rebase ;-)

benpicco commented on sys/shell/commands/sc_gnrc_netif.c:
```diff
@@ -1021,9 +1106,24 @@ static int _netif_set(char *cmd_name, kernel_pid_t iface, char *key, char *value
     }
+    else if (strcmp("deveui", key) == 0) {
+        return _netif_set_addr(iface, NETOPT_ADDRESS_LONG, value);
+    }
```
What's wrong with using `addr_long` for that?

jia200x replied:
technically the DEVEUI is not an address, although in RIOT is modeled using NETOPT_ADDR_LONG.

jia200x replied:
how to proceed? Should I add a NETOPT_DEVEUI option, use `addr_long` or leave it as it is?

benpicco replied:
Well if you feel better with that alias I won't object - but only make it available for `MODULE_GNRC_NETIF_CMD_LORA` to avoid confusion.

jia200x replied:
I think it's better to leave it that way.

At least for the LoRaWAN user, it's usually about "deveui", "appkey" and "appeui". For ABP, "nwskey", "appskey" and the only thing I'm mapping is "addr" to "devaddr" (since I don't think someone would be confused, since it's strictly an address).

I already addressed the `#ifdef` wrap :)

miri64 replied:
The thing is in the review this options were introduced, we came to the conclusion, that `NETOPT_DEVEUI` is the same as `NETOPT_ADDRESS_LONG`, so I would be against introducing an alias. I prefer the duplicate functionality here, as it is what a typical LoRA user would expect in the frontend.

benpicco commented on sys/shell/commands/sc_gnrc_netif.c:
```diff
@@ -1021,9 +1106,24 @@ static int _netif_set(char *cmd_name, kernel_pid_t iface, char *key, char *value
+    }
+    else if (strcmp("nwkskey", key) == 0) {
+        return _netif_set_addr(iface, NETOPT_LORAWAN_NWKSKEY, value);
+    }
```
`#ifdef MODULE_GNRC_NETIF_CMD_LORA`

benpicco commented:
Murdock says you forgot another `#ifdef MODULE_GNRC_NETIF_CMD_LORA` :wink:

jia200x commented:
can I rebase?

jia200x commented:
> Murdock says you forgot another #ifdef MODULE_GNRC_NETIF_CMD_LORA wink

Done! I also fixed a trailing whitespace

benpicco commented:
From my side yes! Makes it easier to review the individual patches.

jia200x commented:
ok, rebased :)

benpicco commented:
And also makes it easier to spot errors in CI 😉

jia200x commented:
> And also makes it easier to spot errors in CI wink

Yup, I'm fixing those errors :)

jia200x commented:
I think I fixed the issues, and rebased to avoid unrelated `gnrc_ipv6_ext_frag` insufficient memory errors

miri64 commented on examples/gnrc_lorawan/main.c:
```diff
@@ -0,0 +1,126 @@
+    .ctx = NULL
+};
+
+static gnrc_netreg_entry_t _entry = GNRC_NETREG_ENTRY_INIT_CB(LORAWAN_PORT, &_cbd);
```
Why not use straight-up `gnrc_pktdump` for that? It basically copies its functionallity (except for only printing the payload).

jia200x replied:
yes, I think we can use that

jia200x replied:
addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/ebe141a79582016423e4b86ba13d678d1b11c82d and https://github.com/RIOT-OS/RIOT/pull/11022/commits/968d64837885c78d6ba0a9921d4263e674462617

miri64 commented on sys/include/net/gnrc/netif/flags.h:
```diff
@@ -129,6 +129,11 @@ enum {
+/**
+ * @brief   A Link Check request was scheduled
+ */
+#define GNRC_NETIF_FLAGS_LINK_CHECK                (0x00020000U)
```
As someone who never worked with LoRa I ask myself immediately: What is a link check? Is this flag reusable for other network interface types?

If it is just LoRa-specific, why not introduce a separate flag field in `gnrc_netif_lorawan_t` and put the flag definition to `sys/include/net/gnrc/netif/lorawan.h`?

jia200x replied:
Yes, I think is LoRaWAN specific. I will move them to the `lorawan.h` header

jia200x replied:
addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/067c753148176139cb420fa87a4361714ce9ac21. I added a commit to revert the changes to the `flags.h` header in https://github.com/RIOT-OS/RIOT/pull/11022/commits/c2052d0a2f695604889131a20fccf4625e09f2d4

miri64 commented on sys/include/net/gnrc/netif/lorawan_base.h:
```diff
@@ -0,0 +1,49 @@
+ * @return  NULL, on error.
+ */
+gnrc_netif_t *gnrc_netif_lorawan_create(char *stack, int stacksize, char priority,
+                                        char *name, netdev_t *dev);
```
Why is this not just in `sys/include/net/gnrc/netif/lorawan.h`?

jia200x replied:
there was a dependency hell at some point, because the `gnrc_netif_t` struct includes the other layers (ipv6, mac layers, etc).

miri64 replied:
I see. Then go ahead as is :-).

miri64 commented on sys/include/net/gnrc/nettype.h:
```diff
@@ -129,6 +129,7 @@ typedef enum {
     /**
      * @}
      */
+    GNRC_NETTYPE_LORAWAN,       /**< Protocol is LoRaWAN */
```
Please put this above `GNRC_NETTYPE_TEST`. Also it is necessary to surround this with a proper guard, otherwise memory is wasted, when `gnrc_lorawan` is not included in the firmware (the value space of `gnrc_nettype_t` decides the size of the `gnrc_netreg` registry at compile-time).

```suggestion
#ifdef MODULE_GNRC_LORAWAN
    GNRC_NETTYPE_LORAWAN,       /**< Protocol is LoRaWAN */
#endif
```

jia200x replied:
addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/610e8250a995b56c695b949ab4593af523188e3d

jia200x commented:
I will piggyback these unrelated errors to this PR

miri64 commented on examples/gnrc_lorawan/Makefile:
```diff
@@ -7,6 +7,7 @@ USEMODULE += gnrc_netdev_default
 USEMODULE += auto_init_gnrc_netif
 USEMODULE += gnrc_lorawan
 USEMODULE += gnrc_netapi_callbacks
```
This is not needed anymore now, right?

jia200x replied:
addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/b87aa2acbd749bee733b29d6ec57844963de4916

miri64 commented on sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:
```diff
@@ -0,0 +1,351 @@
+
+static void _init(gnrc_netif_t *netif)
+{
+    netif->dev->event_callback = _driver_cb;
```
It might be recommendable to add a call to `gnrc_netif_init_default()` here, no that #10485 is merged.

jia200x replied:
you are right, I will.

jia200x replied:
addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/60df21c68451fa5a8374c5bca5bdf5ebf7d8486f.

jia200x commented:
I think I addressed all exposed issues. May I squash?

miri64 commented on sys/include/net/gnrc/netif.h:
```diff
@@ -60,6 +63,12 @@
+/**
+ * @brief   A Link Check request was scheduled
+ */
+#define GNRC_NETIF_LORAWAN_FLAGS_LINK_CHECK                (0x00020000U)
```
Didn't you want to move to `sys/include/net/gnrc/netif/lorawan.h` as we agreed this is LoRaWAN specific (see https://github.com/RIOT-OS/RIOT/pull/11022#discussion_r341591268)?

miri64 replied:
Also `0x00020000U` doesn't fit into the flag field provided by `gnrc_netif_lorawan_t`. It is only `uint8_t`.

miri64 replied:
Otherwise if you *do* want to `gnrc_netif_t::flags` rather than `gnrc_netif_t::lorawan::flags`, it should go back to `sys/include/net/gnrc/netif/flags.h`, but as it is LoRaWAN specific it really shouldn't.

jia200x replied:
oh wait, I thought that was removed... let me check

miri64 commented on sys/include/net/gnrc/netif/lorawan.h:
```diff
@@ -0,0 +1,56 @@
+/**
+ * @brief   A Link Check request was scheduled
+ */
+#define GNRC_NETIF_LORAWAN_FLAGS_LINK_CHECK                (0x00020000U)
```
Still doesn't fit in `gnrc_netif_lorawan_t::flags` ;-). It doesn't have any other flags yet, how about `0x01`?

jia200x replied:
addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/28f641154386b6a021365ca0c4ddb534ff8e1308

miri64 commented on sys/include/net/gnrc/netif/lorawan.h:
```diff
@@ -0,0 +1,56 @@
+    uint8_t port;                           /**< LoRaWAN port for the next transmission */
+    uint8_t deveui[LORAMAC_DEVEUI_LEN];     /**< Device EUI buffer */
+    uint8_t appeui[LORAMAC_APPEUI_LEN];     /**< App EUI buffer */
+    uint8_t appkey[LORAMAC_APPKEY_LEN];     /**< App Key buffer */
```
Why not move those to the other arrayes on top. This might safe some bytes of RAM due to byte alignment.

jia200x replied:
https://github.com/RIOT-OS/RIOT/pull/11022/commits/28f641154386b6a021365ca0c4ddb534ff8e1308

miri64 commented on sys/include/net/gnrc/netif/lorawan.h:
```diff
@@ -0,0 +1,56 @@
+    uint8_t appeui[LORAMAC_APPEUI_LEN];     /**< App EUI buffer */
+    uint8_t appkey[LORAMAC_APPKEY_LEN];     /**< App Key buffer */
+    int ack_req;                            /**< Request ACK in the next transmission */
+    int otaa;                               /**< wether the next transmission is OTAA or not */
```
Do those need to be `int`? Making them `int8_t` would safe some bytes.

jia200x replied:
also in https://github.com/RIOT-OS/RIOT/pull/11022/commits/28f641154386b6a021365ca0c4ddb534ff8e1308 :)

jia200x commented:
@miri64 I also removed some non required `#include` in https://github.com/RIOT-OS/RIOT/pull/11022/commits/0a1a3b61e17e76b982d108178a1530a98682ba5f

jia200x commented:
@miri64 I addressed all GNRC comments. Anything else you might need for ACK'n the GNRC adoption?

jia200x commented:
@smlng @aabadie do you have more comments on this PR?

jia200x commented:
ping @aabadie @smlng

smlng commented on examples/gnrc_lorawan/Makefile:
```diff
@@ -0,0 +1,60 @@
+# `ifconfig`
+
+# OTAA compile time configuration keys
+CFLAGS += -DLORAMAC_APP_KEY_DEFAULT=\{0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\,0x00\}
```
to me this escaped hex notation is not very user friendly, but that's not a blocker for this PR, just something we might want to improve on later.

jia200x replied:
no problem, I can change it :)

jia200x replied:
I just realized this is already defined in RIOT with that format. Can we leave it for later maybe? :)

smlng replied:
yes sure, as said this is non blocking - and rather an inconvenience to users.

jia200x replied:
ok! thanks

smlng commented on sys/include/net/gnrc/lorawan.h:
```diff
@@ -0,0 +1,237 @@
+ * @brief maximum timer drift in percentage
+ *
+ * E.g a value of 0.1 means there's a positive drift of 0.1% (set timeout to
+ * 1000 ms => triggers after 1001 ms)
```
add note that this is only a workaround in RIOT to compensate for bad timers/clocks, right?!

jia200x replied:
https://github.com/RIOT-OS/RIOT/pull/11022/commits/1dec66fe2e7540c7d2e9609c49407a3f1216b404

smlng commented on sys/include/net/gnrc/lorawan/region.h:
```diff
@@ -0,0 +1,71 @@
+};
+
+#define GNRC_LORAWAN_DEFAULT_CHANNELS_NUMOF \
+    (sizeof(gnrc_lorawan_default_channels)/sizeof(uint32_t)) /**< Number of default channels */
```
could use `ARRAY_SIZE` macro here

jia200x replied:
https://github.com/RIOT-OS/RIOT/pull/11022/commits/efbb577458393868cebc84aba388e54828c30eac

smlng commented on sys/include/net/gnrc/lorawan/region.h:
```diff
@@ -0,0 +1,71 @@
+ * @return true if datarate is valid
+ * @return false otherwise
+ */
+int gnrc_lorawan_validate_dr(uint8_t dr);
```
return values do not match function, i.e. if `true` and `false` why not `bool` then, otherwise stick with `int` and explicitly state `1` and `0` here.

jia200x replied:
changed it to bool

jia200x replied:
https://github.com/RIOT-OS/RIOT/pull/11022/commits/9c3a413316973e3417a1eb44225bbfa1f397afd0

smlng commented on sys/include/net/gnrc/netif/lorawan.h:
```diff
@@ -0,0 +1,56 @@
+    uint8_t appskey[LORAMAC_APPSKEY_LEN];   /**< App SKey buffer */
+    uint8_t appkey[LORAMAC_APPKEY_LEN];     /**< App Key buffer */
+    uint8_t deveui[LORAMAC_DEVEUI_LEN];     /**< Device EUI buffer */
+    uint8_t appeui[LORAMAC_APPEUI_LEN];     /**< App EUI buffer */
```
this is big, right? Do we need these 3 in memory, IIRC after joining with OTAA using these we have `nwkskey` and `appskey` for that session, so could we save some RAM here?

jia200x replied:
I can move them to the `sc_netif` files.
We need them in RAM for the shell (e.g `ifconfig set appkey`), but I agree we don't need it in RAM for most LoRaWAN applications

jia200x replied:
after second thoughts, I think it will be hard to remove this now.
In the shell files we only have the interface as context. So, if we don't keep the keys there we cannot do `ifconfig set appkey ...` :(

I propose to do this optimization in a follow up, considering the PR still consumes > 1kb less than the semtech pkg

smlng replied:
To clarify this is more a question bc I'm unsure, so I rely on you expertise. But you seem to agree that these values are not generally required. So I would suggest to move them as you proposed. Because in my lora app I wouldn't use the shell interface to config lora.

smlng replied:
Okay leave as is, and optimise later.

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:
```diff
@@ -0,0 +1,349 @@
+            break;
+        default:
+            gnrc_pktbuf_release(pkt);
+            break;
```
you don't use break on `default` case in other places, be consistent.

jia200x replied:
I added all missing `break` statements in the last push

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,344 @@
+            default:
+                counter = GNRC_LORAWAN_BACKOFF_TIME_3;
+                mac->mlme.backoff_budget = GNRC_LORAWAN_BACKOFF_BUDGET_3;
+                break;
```
you do not use break on `default` case in (most) other usages, stay consistent

jia200x replied:
done!

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,344 @@
+                mac->mlme.backoff_budget = GNRC_LORAWAN_BACKOFF_BUDGET_3;
+                break;
+        }
+
```
nit-picky: remove empty line

jia200x replied:
https://github.com/RIOT-OS/RIOT/pull/11022/commits/83f8846bd22f8c78aa72efef0c1ef25451586eb9

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,344 @@
+    xtimer_set_msg(&mac->mlme.backoff_timer,
+                   GNRC_LORAWAN_BACKOFF_WINDOW_TICK,
+                   &mac->mlme.backoff_msg, thread_getpid());
+
```
and again: remove empty line

jia200x replied:
https://github.com/RIOT-OS/RIOT/pull/11022/commits/83f8846bd22f8c78aa72efef0c1ef25451586eb9

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,344 @@
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+            break;
+        default:
+            break;
```
mhm, maybe you use `break` more often in `default` then I thought 😄

jia200x replied:
addressed :) (see d.g https://github.com/RIOT-OS/RIOT/pull/11022/commits/bda25749eb84ba2310409f887164eb7911ebc4e9)

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,344 @@
+    lorawan_buffer_t buf;
+    _buffer_reset(&buf, fopts, size);
+
+    while (buf.index < buf.size) {
```
I find this a bit hard to understand, bc just looking at this it seems to be a infinite loop. 
Actually its not bc `_mlme_link_check_ans` changes an internal value of `buf`. 

I would rewrite this to something like:

```
int ret = 0;
while ((buf.index < buf.size) && (ret >= 0)) {
    buf.index += ret;
    switch (*(buf.data)) {
        case GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS:
            ret = _mlme_link_check_ans(mac, &buf);
            break;
        default:
            ret = -1;
     }
}
```

then rewrite `_mlme_link_check_ans ` to return `3` as how far the index needs to me moved forward.

What do you think?

jia200x replied:
I replaced this whole snippet to keep check of fopts flow (malformed fopts frame, very unlikely but not out of the specification...)

Besides saving some ROM, the link check command still works:
```
2019-11-21 15:03:27,349 # Iface  4  HWaddr: 26:01:29:77  Frequency: 869524963Hz  BW: 125kHz  SF: 9  CR: 4/5  Link: up 
2019-11-21 15:03:27,356 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 2  Num gateways.: 25 
2019-11-21 15:03:27,359 #           IQ_INVERT 
2019-11-21 15:03:27,362 #           RX_SINGLE OTAA L2-PDU:255 
```

jia200x replied:
@smlng are you OK with the proposed solution? If so, all comments are addressed

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:
```diff
@@ -0,0 +1,124 @@
+
+static uint8_t dr_sf[GNRC_LORAWAN_DATARATES_NUMOF] = { LORA_SF12, LORA_SF11, LORA_SF10, LORA_SF9, LORA_SF8, LORA_SF7 };
+static uint8_t dr_bw[GNRC_LORAWAN_DATARATES_NUMOF] = { LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ };
+
```
coding style, i.e. : break lines

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:
```diff
@@ -0,0 +1,124 @@
+
+void gnrc_lorawan_channels_init(gnrc_lorawan_t *mac)
+{
+    for(unsigned i = 0; i<GNRC_LORAWAN_DEFAULT_CHANNELS_NUMOF; i++) {
```
coding style: spaces around `i < GNRC_L...`

jia200x replied:
I ran uncrustify on this file (see last commit)

tcschmidt commented:
@aabadie it looks like your concerns have been addressed a long time ago - ready to remove your blocking?

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -299,19 +280,24 @@ void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size
+
+        if(pos + ret > size) {
+            return;
+        }
```
IMHO this is redundant with the for-loop condition

jia200x replied:
the problem is, I have to check that it doesn't overflow before executing the fopt payload

jia200x replied:
it's a tri-state (`pos + ret < size`, `pos + rest == size`, `pos + ret > size`). Each one of them has a different behavior.

smlng replied:
ah so this should work then:

```
for(uint8_t pos = 0; pos < (size - 2); pos += ret) {
    switch (fopts[pos]) {
    case GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS:
        ret += 3;
        _mlme_link_check_ans(mac, &fopts[pos + 1]);
        break;
    default:
        return;
    }
}
```

please note: I changed the pointer to be `pos + 1` because you access `p[0]` and `p[1]` in `_mlme_link_check_ans`.

jia200x replied:
That loop doesn't work :(

If size is e.g 2 (the minimum fopt size is 2), then the loop would never be executed

jia200x replied:
done! (https://github.com/RIOT-OS/RIOT/pull/11022/commits/22aa1aee7f670a164f742704f801cd1df6e6ab4c)
```
2019-11-22 13:40:21,070 # Iface  4  HWaddr: 26:01:2B:7D  Frequency: 869524963Hz  BW: 125kHz  SF: 9  CR: 4/5  Link: up 
2019-11-22 13:40:21,077 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 24  Num gateways.: 2 
2019-11-22 13:40:21,080 #           IQ_INVERT 
2019-11-22 13:40:21,083 #           RX_SINGLE OTAA L2-PDU:255 
```

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -299,19 +280,24 @@ void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size
-                    return;
-                }
+                ret += 3;
+                cb = _mlme_link_check_ans;
```
this is hard to get, even though I'm looking at this in the morning, it took me a moment to realise that this just sets a function pointer which is then called later. 

I would suggest to simply call that function here, and not doing this redirection via a function pointer - this is hard to read.

jia200x replied:
as described before, I have to check it doesn't overflow before calling the function :(. I would need to add checks for each function in that case

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -273,16 +259,11 @@ int _fopts_mlme_link_check_req(lorawan_buffer_t *buf)
 }
 
-static int _mlme_link_check_ans(gnrc_lorawan_t *mac, lorawan_buffer_t *fopt)
+static int _mlme_link_check_ans(gnrc_lorawan_t *mac, uint8_t *p)
```
can be void, as you have only a single return value, i.e. `0`, and also not use it below when calling/exec the function

jia200x replied:
done! (https://github.com/RIOT-OS/RIOT/pull/11022/commits/22aa1aee7f670a164f742704f801cd1df6e6ab4c)

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:
```diff
@@ -14,10 +14,13 @@
-static uint8_t dr_sf[GNRC_LORAWAN_DATARATES_NUMOF] = { LORA_SF12, LORA_SF11, LORA_SF10, LORA_SF9, LORA_SF8, LORA_SF7 };
-static uint8_t dr_bw[GNRC_LORAWAN_DATARATES_NUMOF] = { LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ };
+static uint8_t dr_sf[GNRC_LORAWAN_DATARATES_NUMOF] =
+{ LORA_SF12, LORA_SF11, LORA_SF10, LORA_SF9, LORA_SF8, LORA_SF7 };
```
nit pick: looks better to me:

```
static uint8_t dr_sf[GNRC_LORAWAN_DATARATES_NUMOF] = { 
    LORA_SF12, LORA_SF11, LORA_SF10, LORA_SF9, LORA_SF8, LORA_SF7 
};
```

jia200x replied:
this was the output of Uncrustify. Anyway, I can change it if you want

jia200x replied:
as discussed offline, this was Uncrustify =P

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:
```diff
@@ -14,10 +14,13 @@
+static uint8_t dr_sf[GNRC_LORAWAN_DATARATES_NUMOF] =
+{ LORA_SF12, LORA_SF11, LORA_SF10, LORA_SF9, LORA_SF8, LORA_SF7 };
+static uint8_t dr_bw[GNRC_LORAWAN_DATARATES_NUMOF] =
+{ LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ,
```
same here, put 3 values per line

```
static uint8_t dr_bw[GNRC_LORAWAN_DATARATES_NUMOF] = {
    LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ,
    LORA_BW_125_KHZ, LORA_BW_125_KHZ, LORA_BW_125_KHZ
};
```

smlng replied:
looks IMHO nicer and more readable

smlng replied:
ah damn, but that came from the uncrustify, right?!

mhm, still mine looks better 😬

smlng commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -0,0 +1,328 @@
+            return;
+        }
+
+        cb(mac, fopts + pos + 1);
```
last one: use `&fopts[pos + 1]` or (even better (to me)) `&fopts[pos]` and then use `p[1]` and `p[2]` in the other function. The latter would also allow for later changing this to a (nice) struct casting based solution

```
typedef struct {
    uint8_t type;
    uint8_t margin;
    uint8_t ngtws;
} mlme_link_check_ans_t;

mlme_link_check_ans_t *mlca = (mlme_link_check_ans_t *)p;

mlme_confirm->link_req.margin = mlca->margin;
mlme_confirm->link_req.num_gateways = mlca->ngtws;
```

or something like that.

I'd prefer the bracket notation bc its used in the switch, too.

jia200x replied:
done :) (https://github.com/RIOT-OS/RIOT/pull/11022/commits/407d365d7009c8d494f749f568ef2cd70f9f7d0e)

smlng commented:
still need to test again, otherwise looks good to me. Will ACK after testing.

@aabadie how about giving this another shot?

smlng commented:
I tried to use ABP with `example/gnrc_lorawan` but after trying to set appskey or nwkskey the shell hangs, no kernel panic but also no prompt or reaction anymore

jia200x commented:
> I tried to use ABP with example/gnrc_lorawan but after trying to set appskey or nwkskey the shell hangs, no kernel panic but also no prompt or reaction anymore

Weird... did you set the RX2 DR to 3?
Probably it hangs because it's trying to send confirmable messages several times. It should be OK after 30-40 seconds. 
I will give it a look.

smlng commented:
I'll check

smlng commented:
tried again:

- OTAA join works partly, i.e. see join message in TTN console but interface stays down. I guess bc of known timing problem with TTN.
- ABP hangs after setting nwkskey or appskey, i.e. the RIOT shell hangs, no new prompt

smlng commented:
I use BOARD=b-l072z-lrwan1 and the unchanged Makefile for gnrc-lorawan example from this PR, all keys and stuff are entered via shell, i.e. `ifconfig 4 set ...`

jia200x commented:
> ABP hangs after setting nwkskey or appskey, i.e. the RIOT shell hangs, no new prompt

Ok, let me check

jia200x commented:
> ABP hangs after setting nwkskey or appskey, i.e. the RIOT shell hangs, no new prompt
It's not hanging for me :(

```
ifconfig 4 set nwkskey 7442E7AFD925400A9B00C00181734FD8
2019-11-25 11:44:54,352 #  ifconfig 4 set nwkskey 7442E7AFD925400A9B00C00181734FD8
2019-11-25 11:44:54,359 # success: set NwkSKey on interface 4 to 7442E7AFD925400A9B00C00181734FD8
ifconfig 4 set appskey 448FB8A69B74423ECFCA8F53A6BD6F3E
2019-11-25 11:44:57,096 #  ifconfig 4 set appskey 448FB8A69B74423ECFCA8F53A6BD6F3E
2019-11-25 11:44:57,103 # success: set AppSKey on interface 4 to 448FB8A69B74423ECFCA8F53A6BD6F3E
> ifconfig 4 set addr 2601274C
2019-11-25 11:46:33,998 #  ifconfig 4 set addr 2601274C
2019-11-25 11:46:34,003 # success: set (short) address on interface 4 to 2601274C
> ifconfig 4 -otaa
2019-11-25 11:46:54,525 #  ifconfig 4 -otaa
2019-11-25 11:46:54,527 # success: unset option
> ifconfig 4 up
2019-11-25 11:47:08,677 #  ifconfig 4 up
> send 4 hola
2019-11-25 11:47:24,189 #  send 4 hola
2019-11-25 11:48:00,018 # Error sending packet (not joined?)
```
(keys are dummy, so no problem to exposed them)

jia200x commented:
(it's expected to have an error if using confirmable messages and the rx2_dr doesn't match)

jia200x commented:
@smlng I cannot reproduce the error :(

smlng commented:
mhm, I use BUILD_IN_DOCKER and macOS - that are the only specialities I've on my side. I also switched the board, so just in case there is a hardware issue

jia200x commented:
are you using the latest commit? Could you please paste the logs here?

smlng commented:
that's my the HEAD of the PR branch:

```
commit 407d365d7009c8d494f749f568ef2cd70f9f7d0e (HEAD, refs/pull/upstream/11022)
Author: Jose Alamos <jose.alamos@haw-hamburg.de>
Date:   Fri Nov 22 15:42:29 2019 +0100

    fixup! gnrc_lorawan: add initial support for LoRaWAN stack
```

jia200x commented:
could you paste the list of commands that hang the node?

smlng commented:
So I fetched and reset the PR branch, just to be sure. However, still same behaviour:

- OTAA join is send and visible in TTN
- ABP hangs after setting network session key

Command:

```
2019-11-25 15:10:20,042 #  ifconfig
2019-11-25 15:10:20,052 # Iface  4  HWaddr: 00:00:00:00  Frequency: 868299987Hz  BW: 125kHz  SF: 12  CR: 4/5  Link: down 
2019-11-25 15:10:20,059 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 0  Num gateways.: 0 
2019-11-25 15:10:20,061 #           OTAA 
2019-11-25 15:10:20,063 #           L2-PDU:255 
2019-11-25 15:10:20,064 #           
> ifconfig 4 -otaa
2019-11-25 15:10:45,873 #  ifconfig 4 -otaa
2019-11-25 15:10:45,875 # success: unset option
> ifconfig 4 set nwkskey 4A6EE9CD88xxx
2019-11-25 15:11:03,799 #  ifconfig 4 set nwkskey 4A6EE9CD88xxx
2019-11-25 15:11:03,806 # success: set NwkSKey on interface 4 to 4A6EE9CD88xxx
help
```

Key is shortened for security, however it hangs after `ifconfig set`

smlng commented:
just tested again not using MacOS but on Linux and it works - so there seems to be (again) a problem with the RIOT shell and macOS 😭

smlng commented:
@aabadie can you please verify and ack, or at least dismiss your review?

smlng commented:
please squash, too

jia200x commented:
I changed `CONFIG_GNRC_LORAWAN_MIN_SYMBOLS_TIMEOUT` to 30 symbols. With the previous value (50) it prevented the second RX window to be opened at the right time when using DR0.

jia200x commented:
squashed!

jia200x commented:
something happened with TTN in Hamburg :(

I ran some tests with the french TTN (IoT-LAB) and everything works OK! I could join in all DRs

smlng commented:
Retested with the example, still works and TTN in Hamburg is better now, too. My ACK holds!

@aabadie I urge you to (quickly) check your review, we should not be too nit-picky here (I already was 😬) as @jia200x has some more LoRaWAN feature PRs in the pipe.

smlng commented:
@MichelRottleuthner any comments (or ACK, maybe 😄)?

MichelRottleuthner commented:
There is a missing white space at... ...nah just kidding.
No further comments. Since you already ACKed I'll hit this nice green button now ;)

smlng commented:
@aabadie IIRC porting this into a package of its own is still something @jia200x is working on.

smlng commented:
great work, many thanks to @jia200x and everyone reviewing and testing! For the latter: keep on going.

aabadie commented:
Thanks, that's a good work @jia200x!

jia200x commented:
Awesome!! <3 Thanks to everybody!!! I will bake you a cake

tcpipchip commented:
Hi, some news to AU915 or not

thanks

jia200x commented:
hi!

@akshaim is planning to port some regions.
@akshaim do you have plans to take the AU915 too?

tcpipchip commented:
Hi @jia200x 

I am tryng now to port! But not too much experience to understand the code! I am loosing a lot of time :)

@akshaim is the better one to port :)

Thanks

tcpipchip commented:
i am testing on NRF52840
![image](https://user-images.githubusercontent.com/11488460/105198391-606e8600-5b1c-11eb-841a-423b7f08e9e8.png)

jia200x commented:
maybe we could synchronize here! it's indeed interesting to have new regions here. Maybe we could wait for @akshaim regarding plans about this region and see what we can do

jia200x commented:
you can also use the Semtech packet implementation if you need AU915 (see `examples/lorawan`). But this one is not integrated to GNRC.

tcpipchip commented:
Yes! you are correct! We have to wait @akshaim 

Should be better to GNRC :)

thanks @jia200x  @

Are you from Portugal

tcpipchip commented:
other question, must i use the D01, D02 and D03

jia200x commented:
Hi @tcpipchip

> Are you from Portugal

No, but I speak spanish :)

> other question, must i use the D01, D02 and D03

It depends on how far you are from the gateway. The lower the datarate the higher range, but increased Time on Air and energy consumption. Increased ToA means that the node would send data less frequently in order to respect the 1% Duty Cycle restriction

akshaim commented:
> hi!
> 
> @akshaim is planning to port some regions.
> @akshaim do you have plans to take the AU915 too?

Hey @jia200x. Yup, the plan as discussed before is to get ADR tested and working this month and US915 by mid-Feb.

akshaim commented:
> NRF52840

Hey, but this is a BT SoC. Did you use an external radio such as RFM95W ?

tcpipchip commented:
Yes, connected to RFM95W
NRF52840 already tested and Working with Arduino, Micropython and Circuitpython, using LoRaWAN **AU915**
Javascript worked with LoRa
Now i want to test using the gnrc_lorawan
Let me Know when ready to AU915. Used hre in Brazil!
Thanks @akshaim

akshaim commented:
You should be able to use it with [Semtech LoRaMAC](https://github.com/RIOT-OS/RIOT/tree/master/examples/lorawan) for the time being.  Can you please test the [SX1276 driver](https://github.com/RIOT-OS/RIOT/tree/master/tests/driver_sx127x) ?

tcpipchip commented:
do you mean make a new GIT on that github and compile your example

tcpipchip commented:
![image](https://user-images.githubusercontent.com/11488460/105212462-4046c300-5b2c-11eb-9226-a7ba36150ba0.png)

tcpipchip commented:
![image](https://user-images.githubusercontent.com/11488460/105216194-fdd3b500-5b30-11eb-8ea8-22bdba5b3d05.png)

tcpipchip commented:
looks that is reading the register!
About DO0,DO1 and DO2 sx1276 gpios, must i connect then

tcpipchip commented:
but, looks that is not transmiting and receiving! I always try with a LoRa Module LOM204 (WISOL)
![image](https://user-images.githubusercontent.com/11488460/105216456-586d1100-5b31-11eb-9c2f-29d0088a4722.png)

tcpipchip commented:
maybe we have to wait your LoRaWAN version with 915AU

tcpipchip commented:
Hi,
I add the DIOx and...
![image](https://user-images.githubusercontent.com/11488460/105222868-d2a19380-5b39-11eb-92f6-ac5c9526a209.png)
works!

jia200x commented:
> other question, must i use the D01, D02 and D03

Sorry man! I though you meant Data Rates, not DIOs...

Actually you only need to connect DIO0 and DIO1. All the rest are not used so far.

tcpipchip commented:
no problem :)

tcpipchip commented:
now, we have to wait the LoRaWAN :)

akshaim commented:
As you have started a new [thread](https://forum.riot-os.org/t/nrf52840-feather-sx1276/) in forums. Let's move the discussion there.
